### ISSUE TITLE
How to run CoSMIX with SCONE

### ISSUE THREAD

**Issue by SamuelGong** [acsl-technion/cosmix#6]
Hi CoSMIX Team,

I wonder how to run `SUVM` applications with mstore support inside `SCONE`.

In the `readme` I see that `Graphene-SGX` needs to undergo some necessary changes so as to do the same thing. Is this also true for `SCONE`? What is the root cause for the requirement?

Thanks for your attention.

**Comment by shmeni** [comment]
Hi, 

To run with SCONE you can use the compile flag: SDK_BUILD="-DSCONE_BUILD"
Kindly note, we didnt test with newer versions of SCONE since the initial code release so please let us know if you run into trouble. 

SUVM as its currently implemented uses untrusted memory as the backing store. In SCONE we were able to allocate untrusted memory without changes. In Graphene-SGX, however, we added an untrusted memory allocation call since we didnt find a mechanism to allocate untrusted memory from the shielded program.

**Comment by SamuelGong** [comment]
Hey, thanks for your prompt reply! Now I am clear about the answer to my previous questions :)

However, if time allows, could you possible explain to me further about **why we were able to allocate untrusted memory without change in SCONE**? To me, from SCONE's [OSDI '16 paper](https://www.usenix.org/conference/osdi16/technical-sessions/presentation/arnautov), there was no direct hint indicating that one could explicitly allocate untrusted memory. As for untrusted memory, I think the relevant statement should be that

* "In addition, SCONE reduces expensive memory accesses within the enclave by maintaining encrypted application data, such as cached files and network buffers, in non-enclave memory."

and this was achieved by a dedicated secure `ephemeral file system` through its file system shield, as hinted later in the paper. So, **was this the very means you managed to allocate non-enclave memory**?

Looking forward to your help, again!

**Comment by shmeni** [comment]
I'm not affiliated with SCONE so unfortunately, I can't provide details for it. Instead, you can refer to the SCONE documentation to learn more. I believe it's @ https://scontain.com/

To see how we allocated untrusted memory in SCONE for SUVM you can refer to the code. Specifically: https://github.com/acsl-technion/cosmix/blob/master/mstores/suvm/suvm_runtime.cpp#L114

Hope it helps clarify things.

**Comment by SamuelGong** [comment]
Hi CoSMIX Team,

I did not close this issue since I had not found my way out this problem. Please allow me to reopen it.

I now have a toy program `$PWD/suvm/hello.cpp` as follows:

```C++
#include <iostream>
using namespace std;

int main(){
    int length = 1024 * 1024 / sizeof(int);
    int *array;
    array = new int[length];
    for (int i = 0; i < length; i++)
        array[i] = i;
    cout << "here" << endl;
    delete [] array;
}
```

According to your previous advice, I tried to compile it using the flag `-DSCONE_BUILD`, i.e., to define `SCONE_BUILD` to be `1`. Then I tried to run it inside a SCONE container. To be exact, I executed the following commands:

```sh
docker run -v $PWD/suvm:/suvm -dit --name cosmix cosmix
docker exec cosmix bash -c "cd /suvm && g++ -DSCONE_BUILD -o hello hello.cpp && ./hello"

docker run -v $PWD/suvm:/suvm \
-dit --name crosscompiler registry.scontain.com:5050/sconecuratedimages/crosscompilers
docker exec crosscompiler bash -c "cd /suvm && ./hello"
```

However, the result was like

```
3b14dd4a6c8b00ab7be0bdbd56df112f23e9d1a969e52ddd90bbe90204d7e3fd
here
4f0352cdbf6daf92803e366b1c5712d673269f4b5e6c86898efef526e6ad8be4
bash: ./hello: No such file or directory
```

In other words, the generated binary could run in CoSMIX container, but failed in SCONE container. My guess is that the binary did not get compiled properly. Did I miss anything during the compilation?

**Comment by shmeni** [comment]
I meant you should use CoSMIX with the SCONE_BUILD flag, and not g++. let me provide more details and hopefully it'll be clearer.

To compile CoSMIX to support the SCONE containers you can use something like the following (lines 8-10 to compile the pass, the runtime and the mstores): https://github.com/acsl-technion/cosmix/blob/4e67f559a8f065b6367d83be7ee1f3e160437c73/samples/mem_access_benchmark/Makefile#L8  

Then to compile an executable with CoSMIX you should use the compilation steps like (lines 18-23): https://github.com/acsl-technion/cosmix/blob/4e67f559a8f065b6367d83be7ee1f3e160437c73/samples/mem_access_benchmark/Makefile#L18
This will get an object file, which you can finally compile to SCONE using the SCONE compiler instead of simple g++ (or if you choose to use SCONE with host-compiled binaries you can use g++)

Finally, regarding your toy program. I suggest testing suvm with malloc&free instead of new&delete since they are not supported in this version of CoSMIX afaik.

**Comment by SamuelGong** [comment]
Thanks for your help and I can get CoSMIX-compiled binaries run in SCONE now!

---

By the way, can you explain in a high level on what application the `test.c` in `cosmix/samples/mem_access_benchmark` trying to test?

**Comment by SamuelGong** [comment]
Let's move the discussion of `mem_access_benchmark` to #8. Thank you very much for walk me out the original problem!
